content/posts/2020/04/rocksdb/index.eng.md

```diff
@@ -0,0 +1,102 @@
+authors: [seunghun.lee]
+translators: [kidon.seo]
+---
+
+Hello, I'm Seunghun Lee, [Libplanet] developer at Planetarium.
+
+Libplanet provides a storage layer abstraction interface called [`IStore`] and its basic implementation called [`DefaultStore`]. `DefaultStore` had been used to develop [Nine Chronicles] and although it was included as a base in Libplanet and had the upside of being able to use it immediately, there were certainly limitations in terms of performance and storage efficiency.
+
+After reviewing various alternative storage methods, we decided that [RocksDB], a [Key-Value Database] library developed by Facebook, was our best option. We decided to create an `IStore` implementation called [`RocksDBStore`] to use as a backend. In this article, I would like to share our experiences in developing `RocksDBStore`.
+
+
+[Libplanet]: https://libplanet.io/
+[`IStore`]: https://docs.libplanet.io/0.8.0/api/Libplanet.Store.IStore.html
+[`DefaultStore`]: https://docs.libplanet.io/0.8.0/api/Libplanet.Store.DefaultStore.html
+[RocksDB]: https://rocksdb.org/
+[Nine Chronicles]: https://nine-chronicles.com/
+[Key-Value Database]: https://en.wikipedia.org/wiki/Key-value_database
+[`RocksDBStore`]: https://github.com/planetarium/libplanet/blob/master/Libplanet.RocksDBStore/RocksDBStore.cs
+
+## Including Dependent Libraries[^1]
+
+RocksDB relies on other libraries for compression or memory allocation purposes. Unlike [Windows build], for macOS and Linux, RocksDB native libraries must also be installed in the system to use them in the form of dynamic link libraries (*.so* and *.diib*).
+
+In a typical server app, it is natural to install all dependent libraries in your system. This is because the system that runs the server app is usually operated only for that server app. But because we're building an app that serves as blockchain nodes and runs on gamers' systems, it's hard to ask all gamers to install these libraries separately.
+
+So what we came up with was to put the libraries that RocksDB relies on in the game client and distribute them. However, building RocksDB in the form of a dynamic link library without any modifications caused the built RocksDB library to not be able to find the dependent libraries that were included in the game clients.
+
+To resolve this issue, we used a method of modifying [rpath] in the RocksDB dynamic link library file. rpath refers to <q>run-time search path</q>, which is hard-coded within a library file or exe file so the [dynamic linking] [loader] can find the required library in that file. Initially, we considered modifying the rpath when building the RocksDB library, but we eventually decided to modify the rpath in the completed library file because the build script in RocksDB turned out to be more complicated than we thought. Fortunately, with tools called [`install_name_tool`] on macOS and [`patchelf`] on Linux, you can simply modify the rpath to the directory where the current RocksDB library exists.
```

Comment:
```suggestion
To resolve this issue, we used a method of modifying [rpath] in the RocksDB dynamic link library file. rpath refers to <q>run-time search path</q>, which is hard-coded within a library file or executable file so the [dynamic linking] [loader] can find the required library in that file. Initially, we considered modifying the rpath when building the RocksDB library, but we eventually decided to modify the rpath in the completed library file because the build script in RocksDB turned out to be more complicated than we thought. Fortunately, with tools called [`install_name_tool`] on macOS and [`patchelf`] on Linux, you can simply modify the rpath to the directory where the current RocksDB library exists.
```